### PR TITLE
Extract tcp_info for mux tcp bearer

### PR DESCRIPTION
**Pull request by karknu**
Extract and trace TCP rtt, rtt var and congestion window for tcp
sockets.
This is a costly debug trace that only works on linux and requires one
to compile with the flag "tracetcpinfo".

### CODE REVIEW

File: network-mux/src/Network/Mux/TCPInfo.hsc [hunk 1]
```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE CPP                 #-}
+
+#ifdef linux_HOST_OS
```
Comment (coot) [inline network-mux/src/Network/Mux/TCPInfo.hsc:3]: Instead of a top level #ifdef we can conditionally include this module in `other-modules`.

Comment (karknu) [reply network-mux/src/Network/Mux/TCPInfo.hsc:3]: That file defines a `StructTCPInfo` regardless of OS which is used when defining the `MuxTraceTCPInfo`.

Comment (karknu) [reply network-mux/src/Network/Mux/TCPInfo.hsc]: I've changed to a conditional inclusion of TCPInfo in the cabal file and updated the definition of MuxTraceTCPInfo

File: network-mux/src/Network/Mux/Trace.hs [hunk 2]
```diff
@@ -184,4 +194,17 @@ instance Show MuxTrace where
     show (MuxTraceStartedOnDemand mid dir) = printf "Started on demand (%s) in %s" (show mid) (show dir)
     show (MuxTraceTerminating mid dir) = printf "Terminating (%s) in %s" (show mid) (show dir)
     show MuxTraceShutdown = "Mux shutdown"
+#ifdef linux_HOST_OS
+    show (MuxTraceTCPInfo StructTCPInfo { tcpi_snd_mss, tcpi_rcv_mss, tcpi_lost, tcpi_retrans
+                                  , tcpi_rtt, tcpi_rttvar, tcpi_snd_cwnd } len)
```
Comment (coot) [inline network-mux/src/Network/Mux/Trace.hs:199]:
```suggestion
    show (MuxTraceTCPInfo StructTCPInfo
            { tcpi_snd_mss, tcpi_rcv_mss, tcpi_lost, tcpi_retrans
            , tcpi_rtt, tcpi_rttvar, tcpi_snd_cwnd }
            len)
```

File: network-mux/src/Network/Mux/Trace.hs [hunk 3]
```diff
@@ -147,6 +152,11 @@ data MuxTrace =
     | MuxTraceStartedOnDemand MiniProtocolNum MiniProtocolDir
     | MuxTraceTerminating MiniProtocolNum MiniProtocolDir
     | MuxTraceShutdown
+#ifdef linux_HOST_OS
+    | MuxTraceTCPInfo StructTCPInfo Word16
+#else
+    | MuxTraceTCPInfo Word16
+#endif
```
Comment (coot) [inline network-mux/src/Network/Mux/Trace.hs:159]: I haven't though about this, but with your original solution one was able to unify this constructor to
```
    | MuxTraceTCPInfo StructTCPInfo Word16
```

You can leave it as is, and I'll fix it next week (by providing two modules `Network.Mux.TCPInfo.Linux` and a public `Network.Mux.TCPInfo`).

**Review by coot — APPROVED** [review]
LGTM

File: network-mux/src/Network/Mux/Trace.hs (hunk 3 above)
Comment (karknu) [reply network-mux/src/Network/Mux/Trace.hs:159]: Sounds good.

**Pull request merged**